src/identity/configuration/CachedJwkGenerator.ts

```diff
@@ -0,0 +1,74 @@
+import { createPublicKey } from 'crypto';
+import type { KeyObject } from 'crypto';
+import { exportJWK, generateKeyPair, importJWK } from 'jose';
+import type { AsymmetricSigningAlgorithm } from 'oidc-provider';
+import type { KeyValueStorage } from '../../storage/keyvalue/KeyValueStorage';
+import type { AlgJwk, JwkGenerator } from './JwkGenerator';
+
+/**
+ * Generates a key pair once and then caches it using both an internal variable a {@link KeyValueStorage}.
+ * The storage makes sure the keys remain the same between server restarts,
+ * while the internet variable makes it so the storage doesn't have to be accessed every time a key is needed.
+ *
+ * Only the private key is stored in the internal storage, using the `storageKey` parameter.
+ * The public key is determined based on the private key and then also stored in memory.
+ */
+export class CachedJwkGenerator implements JwkGenerator {
+  public readonly alg: AsymmetricSigningAlgorithm;
+
+  private readonly key: string;
+  private readonly storage: KeyValueStorage<string, AlgJwk>;
+
+  private privateJwk?: AlgJwk;
+  private publicJwk?: AlgJwk;
+
+  public constructor(alg: AsymmetricSigningAlgorithm, storageKey: string, storage: KeyValueStorage<string, AlgJwk>) {
+    this.alg = alg;
+    this.key = storageKey;
+    this.storage = storage;
+  }
+
+  public async getPrivateKey(): Promise<AlgJwk> {
+    if (this.privateJwk) {
+      return this.privateJwk;
+    }
+
+    const jwk = await this.storage.get(this.key);
+    if (jwk) {
+      this.privateJwk = jwk;
+      return jwk;
+    }
+
+    const { privateKey } = await generateKeyPair(this.alg);
+
+    // Make sure the JWK is a plain node object for storage
+    const privateJwk = { ...await exportJWK(privateKey) } as AlgJwk;
+    privateJwk.alg = this.alg;
+
+    await this.storage.set(this.key, privateJwk);
+    this.privateJwk = privateJwk;
+    return privateJwk;
+  }
+
+  public async getPublicKey(): Promise<AlgJwk> {
+    if (this.publicJwk) {
+      return this.publicJwk;
+    }
+
+    const privateJwk = await this.getPrivateKey();
+
+    // The main reason we generate the public key from the private key is, so we don't have to store it.
+    // This allows our storage to not break previous versions where we only used the private key.
+    // In practice this results in the same key.
+    const privateKey = await importJWK(privateJwk);
+    const publicKey = createPublicKey(privateKey as KeyObject);
+
+    const publicJwk = { ...await exportJWK(publicKey) } as AlgJwk;
+    // These fields get lost during the above proces
+    publicJwk.alg = privateJwk.alg;
+
+    this.publicJwk = publicJwk;
+
```

Comment:
```suggestion
    if (!this.publicJwk) {
      // …
      this.publicJwk = publicJwk;
    }
```

Reply:
Interesting as I'm much more fan of the original. I generally try to avoid large blocks of code in a single if-statement as I find it harder to then see exactly where the if-statement ends (mostly relevent if there are several blocks though).

Reply:
Makes sense; I typically optimize for a single return statement. But the early return also makes sense (I just have that it's 4 lines instead of the needed 2 😅 and if it were Ruby, we'd just inline `return self.publikJwk unless !self.publikJwk`).